File: packages/conversationmodel/conversation_types.ts
export enum ConversationRoleEnum {
	system = 'system',
	user = 'user',
	assistant = 'assistant',
	function = 'function',
	feedback = 'feedback',
}
export interface ConversationMessage {
	id: string;
	createdAt?: Date;
	role: ConversationRoleEnum;
	content: string;
	timestamp?: string;
	name?: string;
	details?: string;
}

export interface Conversation {
	id: string;
	title: string;
	createdAt: Date;
	modifiedAt: Date;
	messages: ConversationMessage[];
}

export interface IConversationAPI {
	saveConversation: (conversation: Conversation) => Promise<void>;
	deleteConversation: (id: string, title: string) => Promise<void>;
	getConversation: (id: string, title: string) => Promise<Conversation | null>;
	listConversations: (
		token?: string
	) => Promise<{ conversations: { id: string; title: string }[]; nextToken?: string }>;
	addMessageToConversation(id: string, title: string, newMessage: ConversationMessage): Promise<void>;
}

File: pkg/conversationstore/spec/conversation_types.go
package spec

import (
	"time"
)

// ConversationRoleEnum represents the role of a participant in a conversation.
type ConversationRoleEnum string

const (
	ConversationRoleSystem    ConversationRoleEnum = "system"
	ConversationRoleUser      ConversationRoleEnum = "user"
	ConversationRoleAssistant ConversationRoleEnum = "assistant"
	ConversationRoleFunction  ConversationRoleEnum = "function"
	ConversationRoleFeedback  ConversationRoleEnum = "feedback"
)

// ConversationMessage represents a message in a conversation.
type ConversationMessage struct {
	ID        string               `json:"id"`
	CreatedAt *time.Time           `json:"createdAt,omitempty"`
	Role      ConversationRoleEnum `json:"role"`
	Content   string               `json:"content"`
	Name      *string              `json:"name,omitempty"`
	Details   *string              `json:"details,omitempty"`
}

// Conversation represents a conversation with messages.
type Conversation struct {
	ID         string                `json:"id"`
	Title      string                `json:"title"`
	CreatedAt  time.Time             `json:"createdAt"`
	ModifiedAt time.Time             `json:"modifiedAt"`
	Messages   []ConversationMessage `json:"messages"`
}
